lib/AST/RequirementMachine/RewriteSystem.cpp

```diff
@@ -705,10 +718,99 @@ void RewriteSystem::verifyRewriteRules(ValidityPolicy policy) const {
 #undef ASSERT_RULE
 }
 
+std::vector<RequirementError>
+RewriteSystem::getRedundantRequirements() {
+  // The resulting set of redundant requirements.
+  std::vector<RequirementError> redundantRequirements;
+
+  // Map redundant rule IDs to their rewrite path for easy access
+  // in the `isRedundantRule` lambda.
+  llvm::SmallDenseMap<unsigned, RewritePath> redundantRules;
+  for (auto &pair : RedundantRules)
+    redundantRules[pair.first] = pair.second;
+
+  // Collect all rule IDs for each unique requirement ID.
+  llvm::SmallDenseMap<unsigned, llvm::SmallDenseSet<unsigned, 2>>
+      rulesPerRequirement;
+
+  // Collect non-explicit requirements that are not redundant.
+  llvm::SmallDenseSet<unsigned, 2> impliedRequirements;
+
+  for (unsigned ruleID : indices(getRules())) {
+    auto &rule = getRules()[ruleID];
+
+    if (!rule.isExplicit() && !rule.isPermanent() && !rule.isRedundant() &&
```

Comment:
isExplicit() doesn't imply getRequirementID() is not None though does it? Might be best to just use getRequirementID().hasValue() here